### PR TITLE
add kcp workload cordon/uncordon/drain

### LINKED ISSUE #525: kubectl kcp workload cordon|drain [cluster]
**Is your feature request related to a problem? Please describe.**

As a pcluster admin, I want to easily cordon, uncordon and drain clusters using the kcp CLI plugin

**Describe the solution you'd like**

Add CLI surfaces to the `kubectl kcp` CLI plugin to cordon and drain clusters:

- `kubectl kcp workload cordon [cluster]`
- `kubectl kcp workload drain [cluster]`
- `kubectl kcp workload uncordon [cluster]`

**Describe alternatives you've considered**

People can just patch the Cluster resource today, but that's a bit clunkier.

**Additional context**

The behavior of the CLI should mirror the existing `kubectl` surface for (un-)cordoning and draining nodes in a cluster: https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/

### PR DESCRIPTION
**Pull request by chrisahl**
Signed-off-by: Chris Ahl <cahl@redhat.com>

## Summary
Add kcp cli cmds for workload cordon/drain/uncordon

## Related issue(s)
Fixes #525

### CODE REVIEW

**Comment by ncdc** [comment]
@chrisahl thanks! If this closes out the related issue, please make sure in the description it says `Fixes #525`

File: pkg/cliplugins/workload/plugin/workload.go
Comment (ncdc) [inline pkg/cliplugins/workload/plugin/workload.go:82]: Mind switching this to a Patch call instead? Example: https://github.com/kcp-dev/kcp/blob/3f393f5389079c194625f6cd357b847553f11d4a/pkg/cliplugins/workload/plugin/sync.go#L191-L216

File: pkg/cliplugins/workload/plugin/workload.go
Comment (ncdc) [inline pkg/cliplugins/workload/plugin/workload.go:116]: Ditto re Patch

File: pkg/cliplugins/workload/plugin/workload.go
Comment (ncdc) [inline pkg/cliplugins/workload/plugin/workload.go:114]: UX wise, would you want to reject drain if unschedulable is unset, or automatically set it to true?

File: pkg/cliplugins/workload/plugin/workload.go
Comment (ncdc) [inline pkg/cliplugins/workload/plugin/workload.go:111]: Do you think it's useful to let the user optionally specify a time in the future, e.g "drain in 30 minutes"?

**Comment by chrisahl** [comment]
So this is still WIP and I was not yet expecting you to review.  The `drain` is currently async and for nodes they have it sync.  Still a bunch to investigate on `drain`, (and maybe un/cordon too) but I did want to at least open the PR in draft to be sure I could get a PR pushed. 

I will incorporate all the comments already provided (Thanks!) and will have to get used to how your team does PR flow.  Seems all teams do something slightly different so a learning curve on that.

**Comment by ncdc** [comment]
> So this is still WIP and I was not yet expecting you to review

Sorry about that, was trying to provide some fast feedback 😄. I'll hold off and wait for you to let us know when it's ready. Thanks for the PR!

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go:114]: Based on what I see in kubectl, `drain` causes unschedulable to be set and `evictAfter` to be set

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go]: Changed to patches

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go]: Changed to patches

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go:141]: At a future point, if a customer really needs it, a time delay option might be able to be added.  But right now, I don't see if being needed.   There are various aspects of drain for pod that cannot be done for workloadcluster.  Most importantly the ability to have it be sync so that when it returns, the drain is complete.  I think there would need to be additional status added to workloadcluster in order for drain to know when to return.

File: pkg/cliplugins/workload/plugin/workload.go
Comment (stevekuznetsov) [inline pkg/cliplugins/workload/plugin/workload.go:97]: If I understand this correctly, this will encode both the UID and resourceVersion in the patch, which will cause this command to fail when either change while it's running. For UID, that seems like what we want. What do we expect for UX on the resourceVersion? Do we want this command to fail if it races with some other status update or something? If not, we should put this into a retry loop or remove the resourceVersion from the patch.

Comment (stevekuznetsov) [reply pkg/cliplugins/workload/plugin/workload.go:97]: [Here](https://github.com/kcp-dev/kcp/blob/69aed93556a6729fcc8c5e11eed458b11603e0f8/pkg/reconciler/tenancy/clusterworkspaceshard/clusterworkspaceshard_controller.go#L165-L171) is an example of building an explicit new object to diff against. You'd just want to do it with the Spec.

Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go]: @stevekuznetsov I have submitted a change based on your comments

File: pkg/cliplugins/workload/plugin/workload.go
Comment (sttts) [inline pkg/cliplugins/workload/plugin/workload.go:118]: isn't a jsonpatch basically a one-liner for the same 20 merge patch generation code here?

File: pkg/cliplugins/workload/plugin/workload.go
Comment (sttts) [inline pkg/cliplugins/workload/plugin/workload.go:137]: wrong comment?

File: pkg/cliplugins/workload/cmd/cmd.go
Comment (sttts) [inline pkg/cliplugins/workload/cmd/cmd.go:184]: wouldn't a `--evict-after` flag make sense to override the "now" timestamp?

File: pkg/cliplugins/workload/cmd/cmd.go
Comment (sttts) [inline pkg/cliplugins/workload/cmd/cmd.go:45]: more like "Start draining". This command does not wait. We might want to add waiting in the future vi a `--wait`

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go:137]: fixing

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go:118]: Not sure I understand your comment.  Previous comment by @stevekuznetsov indicated to change from Update to Patch and gave an example that I followed.

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go:100]: The patch did not work if that was not there.
```
k kcp workload uncordon cahl-aws-4-10-5         
error: failed to update WorkloadCluster cahl-aws-4-10-5: workloadclusters.workload.kcp.dev "cahl-aws-4-10-5" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update
```

Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go:100]: And this follows the example you linked to.

Comment (ncdc) [reply pkg/cliplugins/workload/plugin/workload.go:100]: You'd need to clear the resource version in the old object before converting it to json. Then you can exclude it from the new object too

File: pkg/cliplugins/workload/plugin/workload.go
Comment (ncdc) [reply pkg/cliplugins/workload/plugin/workload.go:118]: Stefan is suggesting to manually create the patch bytes

File: pkg/cliplugins/workload/cmd/cmd.go
Comment (chrisahl) [reply pkg/cliplugins/workload/cmd/cmd.go:184]: an evict-after option could be added later if really needed but I do not see it as being needed to start with.  kubectl node drain does not offer such an option

File: pkg/cliplugins/workload/plugin/workload.go
Comment (chrisahl) [reply pkg/cliplugins/workload/plugin/workload.go]: I have changed the way the patch is built. See if it is what you are looking for (there are several ways to build patches)

**Comment by chrisahl** [comment]
@sttts @ncdc @stevekuznetsov Can I get another review?

File: pkg/cliplugins/workload/cmd/cmd.go
Comment (sttts) [reply pkg/cliplugins/workload/cmd/cmd.go:184]: Agree.

File: pkg/cliplugins/workload/cmd/cmd.go
Comment (sttts) [reply pkg/cliplugins/workload/cmd/cmd.go]: +1

**Review by ncdc — APPROVED** [review]

**Pull request merged**
